### PR TITLE
perf: Improve internal type discovery and getPropertyInternal

### PR DESCRIPTION
**Pull request by obastemur**
Acme AIR LTO gain ~2%

- Fast path for IsNull / IsUndefined (in case the argument is
recyclableObject, or we know for sure object's context). Especially
NullCheck was pretty common within the internal loops

- Introduce internal GetProperty calls by introducing
GetPropertyNoCache that serves only for RecyclableObject+ and when info ==
nullptr.

- Improve logic in related places. (i.e. use less GetTypeId calls,
GetPropertyName calls...)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -155,7 +155,7 @@ if(CC_TARGETS_AMD64)
     add_definitions(-D_M_X64_OR_ARM64)
     add_compile_options(-msse4.2)
 
-    set(CAN_BUILD_WABT 1)
+    #set(CAN_BUILD_WABT 1)
```
Comment (digitalinfinity) [inline CMakeLists.txt]: Why this change? (cc @Cellule)

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 2]
```diff
@@ -10479,11 +10502,43 @@ namespace Js
...
     }
 
+    BOOL JavascriptOperators::IsNull(Var instance, ScriptContext* scriptContext)
+    {
+        return JavascriptOperators::IsNull(instance, scriptContext->GetLibrary());
+    }
+
+    BOOL JavascriptOperators::IsNull(Var instance, JavascriptLibrary* library)
+    {
+        Assert(!RecyclableObject::Is(instance) ? 1 : ((RecyclableObject*)instance)->GetScriptContext()->GetLibrary() == library );
```
Comment (digitalinfinity) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Use TRUE/FALSE instead of hardcoding constants here

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 3]
```diff
@@ -10494,20 +10549,27 @@ namespace Js
...
 
     BOOL JavascriptOperators::IsUndefinedObject(Var instance, ScriptContext *scriptContext)
     {
-        return JavascriptOperators::IsUndefinedObject(instance, scriptContext->GetLibrary()->GetUndefined());
+        return JavascriptOperators::IsUndefinedObject(instance, scriptContext->GetLibrary());
     }
 
     BOOL JavascriptOperators::IsUndefinedObject(Var instance, JavascriptLibrary* library)
     {
+        Assert(!RecyclableObject::Is(instance) ? 1 : ((RecyclableObject*)instance)->GetScriptContext()->GetLibrary() == library );
```
Comment (digitalinfinity) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Use TRUE/FALSE instead of magic numbers

File: CMakeLists.txt (hunk 1 above)
Comment (Cellule) [reply CMakeLists.txt]: Indeed, we need wabt for some of our test on xplat (agreed very few passes).
What we need to do is not set this on release builds, but leave it on test/debug builds

Comment (obastemur) [reply CMakeLists.txt]: no worries! was experimenting with the CI. libwabt is no longer part of the binary for xplat release builds

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 4]
```diff
@@ -1679,6 +1679,34 @@ namespace Js
         return GetProperty_Internal<false>(instance, RecyclableObject::FromVar(instance), true, propertyId, value, requestContext, info);
     }
 
+    BOOL JavascriptOperators::GetProperty_InternalSimple(Var instance, RecyclableObject* object, PropertyId propertyId, Var* value, ScriptContext* requestContext)
```
Comment (Penguinwizzard) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Could you SAL-annotate the outparam?

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 5]
```diff
@@ -10706,6 +10768,28 @@ namespace Js
...
     }
 
+    Var JavascriptOperators::GetPropertyNoCache(RecyclableObject* instance, PropertyId propertyId, ScriptContext* requestContext)
+    {
+        return JavascriptOperators::GetPropertyNoCache(instance, instance, propertyId, requestContext);
+    }
+
+    Var JavascriptOperators::GetPropertyNoCache(Var instance, RecyclableObject* propertyObject, PropertyId propertyId, ScriptContext* requestContext)
+    {
+        Var value;
```
Comment (Penguinwizzard) [inline lib/Runtime/Language/JavascriptOperators.cpp]: nit: I see that the current implementation of GetProperty_InternalSimple definitely sets value; however, could you initialize it to nullptr here?

**Comment by obastemur** [comment]
@Cellule after rebasing to latest r1.7, I started to get wasm related CI errors (Windows) for this PR. Could you please take a look what may cause?

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 6]
```diff
@@ -10494,20 +10550,27 @@ namespace Js
...
 
     BOOL JavascriptOperators::IsUndefinedObject(Var instance, ScriptContext *scriptContext)
     {
-        return JavascriptOperators::IsUndefinedObject(instance, scriptContext->GetLibrary()->GetUndefined());
+        return JavascriptOperators::IsUndefinedObject(instance, scriptContext->GetLibrary());
     }
 
     BOOL JavascriptOperators::IsUndefinedObject(Var instance, JavascriptLibrary* library)
     {
+        Assert(!RecyclableObject::Is(instance) ? TRUE : ((RecyclableObject*)instance)->GetScriptContext()->GetLibrary() == library );
```
Comment (Cellule) [inline lib/Runtime/Language/JavascriptOperators.cpp]: From what I understand, this Assert is wrong in Cross Context scenario.
That wasm test just happen to exercice that code path, but it's not a wasm related bug.

**Comment by obastemur** [comment]
@dotnet-bot test Ubuntu static_ubuntu_linux_test please
